[Cocoa] Continue fine-tuning fixed-position page sampling heuristics

whsieh opened the pull request:
#### 48d3a2b6aeb68c595ab0a00d54828145095e09c6
<pre>
[Cocoa] Continue fine-tuning fixed-position page sampling heuristics
<a href="https://bugs.webkit.org/show_bug.cgi?id=289573">https://bugs.webkit.org/show_bug.cgi?id=289573</a>
<a href="https://rdar.apple.com/146801107">rdar://146801107</a>

Reviewed by Abrar Rahman Protyasha.

Make various adjustments to the page color sampling heuristic. See below for more details.

* LayoutTests/fast/page-color-sampling/color-sampling-ignores-non-fixed-content-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/color-sampling-ignores-non-fixed-content.html: Added.

Add a test to verify that sampling doesn&apos;t include content behind the fixed-position container, if
the fixed-position container is partially transparent.

* LayoutTests/fast/page-color-sampling/color-sampling-ignores-text-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/color-sampling-ignores-text.html: Added.

Add a test to verify that text painting is skipped for the purposes of color sampling. To check
this, we include lots of white text inside an orange fixed-position container (with a fairly large
font size, and also decreased `letter-spacing` to maximize the likelihood of sampling the text). At
various offsets relative to the top of the viewport, the sampled color should always remain orange.

* LayoutTests/fast/page-color-sampling/color-sampling-includes-subframes-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/color-sampling-includes-subframes.html: Added.

Add a test to verify that content inside subframes in fixed-position containers are included in the
sampled color.

* Source/WebCore/page/FrameSnapshotting.cpp:
(WebCore::snapshotFrameRectWithClip):
* Source/WebCore/page/FrameSnapshotting.h:

Make several changes to the frame snapshotting flags:

•   `ExcludeReplacedContent`: rename this to `ExcludeReplacedContentExceptForIFrames`. Adjust
    `RenderReplaced::shouldPaint` below to avoid bailing for subframes.

•   `ExcludeText`: add a flag to exclude text from being painted in the snapshot. This is passed
    down as a `PaintBehavior` flag, and eventually checked in `TextBoxPainter::paint`.

•   `FixedAndStickyLayersOnly`: add a flag to limit painting to content in fixed and sticky layers,
    or layers underneath fixed layers. Add a FIXME to include layers inside sticky layers, once we
    have the ability to (quickly) check if we have a sticky ancestor.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::willPaintContents):
* Source/WebCore/page/PageColorSampler.cpp:
(WebCore::PageColorSampler::predominantColor):

If the sampled color is close to being totally transparent, revert to returning an invalid color.

* Source/WebCore/page/TextIndicator.cpp:
(WebCore::snapshotOptionsForTextIndicatorOptions):
* Source/WebCore/rendering/PaintPhase.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::paintForegroundForFragments):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter::paint):

Canonical link: <a href="https://commits.webkit.org/292000@main">https://commits.webkit.org/292000@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7e957a4aec808c798f4acb9b3823c11ff58d3c5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4047 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99675 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45151 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96705 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22667 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72203 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29507 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97657 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10813 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85442 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52537 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10506 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3150 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44488 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3250 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101717 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21687 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15819 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81197 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21935 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81471 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80579 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20115 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25140 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2532 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14904 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21666 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26786 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21336 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24801 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23074 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/PageColorSampler.cpp:
```diff
@@ -306,6 +308,11 @@ Color PageColorSampler::predominantColor(Page& page, const LayoutRect& absoluteR
     if (!pixelBuffer)
         return { };
 
+    auto isNearlyTransparent = [&](const Color& color) {
```
no need to capture anything
```suggestion
    auto isNearlyTransparent = [](const Color& color) {
```

whsieh replied:
Fixed — thanks for the review!

whsieh committed:
[Cocoa] Continue fine-tuning fixed-position page sampling heuristics
https://bugs.webkit.org/show_bug.cgi?id=289573
rdar://146801107

Reviewed by Abrar Rahman Protyasha.

Make various adjustments to the page color sampling heuristic. See below for more details.

* LayoutTests/fast/page-color-sampling/color-sampling-ignores-non-fixed-content-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/color-sampling-ignores-non-fixed-content.html: Added.

Add a test to verify that sampling doesn't include content behind the fixed-position container, if
the fixed-position container is partially transparent.

* LayoutTests/fast/page-color-sampling/color-sampling-ignores-text-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/color-sampling-ignores-text.html: Added.

Add a test to verify that text painting is skipped for the purposes of color sampling. To check
this, we include lots of white text inside an orange fixed-position container (with a fairly large
font size, and also decreased `letter-spacing` to maximize the likelihood of sampling the text). At
various offsets relative to the top of the viewport, the sampled color should always remain orange.

* LayoutTests/fast/page-color-sampling/color-sampling-includes-subframes-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/color-sampling-includes-subframes.html: Added.

Add a test to verify that content inside subframes in fixed-position containers are included in the
sampled color.

* Source/WebCore/page/FrameSnapshotting.cpp:
(WebCore::snapshotFrameRectWithClip):
* Source/WebCore/page/FrameSnapshotting.h:

Make several changes to the frame snapshotting flags:

•   `ExcludeReplacedContent`: rename this to `ExcludeReplacedContentExceptForIFrames`. Adjust
    `RenderReplaced::shouldPaint` below to avoid bailing for subframes.

•   `ExcludeText`: add a flag to exclude text from being painted in the snapshot. This is passed
    down as a `PaintBehavior` flag, and eventually checked in `TextBoxPainter::paint`.

•   `FixedAndStickyLayersOnly`: add a flag to limit painting to content in fixed and sticky layers,
    or layers underneath fixed layers. Add a FIXME to include layers inside sticky layers, once we
    have the ability to (quickly) check if we have a sticky ancestor.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::willPaintContents):
* Source/WebCore/page/PageColorSampler.cpp:
(WebCore::PageColorSampler::predominantColor):

If the sampled color is close to being totally transparent, revert to returning an invalid color.

* Source/WebCore/page/TextIndicator.cpp:
(WebCore::snapshotOptionsForTextIndicatorOptions):
* Source/WebCore/rendering/PaintPhase.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::paintForegroundForFragments):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter::paint):

Canonical link: https://commits.webkit.org/292000@main
